pkg/io/video/scale.go

```diff
@@ -0,0 +1,151 @@
+					imgScaled = image.NewRGBA(rect)
+					if s, ok := scaler.(*draw.Kernel); ok {
+						// *draw.Kernel has optimized version for size fixed scaling
+						dr := rect
+						sr := v.Bounds()
+						scalers[0] = s.NewScaler(dr.Dx(), dr.Dy(), sr.Dx(), sr.Dy())
+					} else {
+						scalers[0] = scaler
+					}
+				}
+				dst := imgScaled.(*image.RGBA)
+				scalers[0].Scale(dst, rect, img, img.Bounds(), draw.Over, nil)
+
+			case *image.YCbCr:
+				if ycbcrNeedRealloc(v, imgScaled) {
+					i := image.NewYCbCr(rect, v.SubsampleRatio)
+					imgScaled = i
+					*planes[0].dst = image.Gray{Pix: i.Y, Stride: i.YStride, Rect: rect}
+					*planes[1].dst = image.Gray{Pix: i.Cr, Stride: i.CStride, Rect: rect}
+					*planes[2].dst = image.Gray{Pix: i.Cb, Stride: i.CStride, Rect: rect}
+
+					if s, ok := scaler.(*draw.Kernel); ok {
+						// *draw.Kernel has optimized version for size fixed scaling
+						dr := rect
+						sr := v.Bounds()
+						scalers[0] = s.NewScaler(dr.Dx(), dr.Dy(), sr.Dx(), sr.Dy())
+						switch v.SubsampleRatio {
+						case image.YCbCrSubsampleRatio444:
+							scalers[1] = s.NewScaler(dr.Dx(), dr.Dy(), sr.Dx(), sr.Dy())
+							scalers[2] = s.NewScaler(dr.Dx(), dr.Dy(), sr.Dx(), sr.Dy())
+						case image.YCbCrSubsampleRatio422:
+							scalers[1] = s.NewScaler(dr.Dx()/2, dr.Dy(), sr.Dx()/2, sr.Dy())
+							scalers[2] = s.NewScaler(dr.Dx()/2, dr.Dy(), sr.Dx()/2, sr.Dy())
+						case image.YCbCrSubsampleRatio420:
+							scalers[1] = s.NewScaler(dr.Dx()/2, dr.Dy()/2, sr.Dx()/2, sr.Dy()/2)
+							scalers[2] = s.NewScaler(dr.Dx()/2, dr.Dy()/2, sr.Dx()/2, sr.Dy()/2)
+						}
+					} else {
+						for i := range scalers {
+							scalers[i] = scaler
+						}
+					}
+
+					switch v.SubsampleRatio {
+					case image.YCbCrSubsampleRatio444:
+					case image.YCbCrSubsampleRatio422:
+						planes[1].dst.Rect.Max.X /= 2
+						planes[2].dst.Rect.Max.X /= 2
+					case image.YCbCrSubsampleRatio420:
+						planes[1].dst.Rect.Max.X /= 2
+						planes[1].dst.Rect.Max.Y /= 2
+						planes[2].dst.Rect.Max.X /= 2
+						planes[2].dst.Rect.Max.Y /= 2
+					}
+				}
+				// Scale each plane
+				*planes[0].src = image.Gray{Pix: v.Y, Stride: v.YStride, Rect: v.Rect}
+				*planes[1].src = image.Gray{Pix: v.Cr, Stride: v.CStride, Rect: v.Rect}
+				*planes[2].src = image.Gray{Pix: v.Cb, Stride: v.CStride, Rect: v.Rect}
+				switch v.SubsampleRatio {
+				case image.YCbCrSubsampleRatio444:
+				case image.YCbCrSubsampleRatio422:
+					planes[1].src.Rect.Max.X /= 2
+					planes[2].src.Rect.Max.X /= 2
+				case image.YCbCrSubsampleRatio420:
+					planes[1].src.Rect.Max.X /= 2
+					planes[1].src.Rect.Max.Y /= 2
+					planes[2].src.Rect.Max.X /= 2
+					planes[2].src.Rect.Max.Y /= 2
+				}
+				// TODO: add option to enable multi threading
+				var wg sync.WaitGroup
+				wg.Add(3)
+				for i, p := range planes {
+					i, p := i, p
+					go func() {
+						scalers[i].Scale(p.dst, p.dst.Rect, p.src, p.src.Rect, draw.Over, nil)
```

Comment:
Scaling Gray image using `x/image/draw` is 10 times slower than RGBA:thinking:

Reply:
Maybe due to Gray->RGBA->Gray color transform?

Reply:
Scaling 1080p to 360p takes:
- YCbCr as 3 `image.Gray`: 60 ms
- YCbCr wrapped by RGBA like image interface: 20ms
- RGBA: 2ms

(`x/image/draw` has optimized implementation only for RGBA to RGBA)

Reply:
1080p RGBA to I444 is now taking 150ms/frame.
After merging #54, screencast example will scale RGBA image, and transform color of downsampled image. So, the performance will be significantly improved.